test/helpers/TaskQueue.spec.js

```diff
@@ -0,0 +1,145 @@
+      assert.equal(count, 5, 'Wrong number of tasks completed');
+    }
+  });
+
+  it('Does not introduce unnecessary delay', async function () {
+    // Create queue with long delay interval
+    const queue = new TaskQueue(40);
+    // Wait 1 ms, so that we are definitely after beginning of first interval
+    await new Promise((r) => { setTimeout(r, 1); });
+
+    const start = new Date().getTime();
+    // Add task to queue
+    const result = await queue.add({ task: () => { return 42; } });
+    const finish = new Date().getTime();
+
+    // Ensure we did not wait to complete task
+    assert(finish - start < 5, 'Queue introduced unnecessary delay');
+    assert.equal(result, 42, 'Task did not complete properly');
+  });
+
+  it('Pauses the queue until a certain time', async function () {
+    // Create queue to test, with short interval
+    const queue = new TaskQueue(1);
+    // Pause queue for 10 ms
+    const start = new Date().getTime();
+    const pauseEnds = new Date(start + 10);
+    queue.pauseUntil(pauseEnds);
+    // Add a task that returns current time to queue
+    const executionTime = await queue.add({
+      task: () => { return Date.now(); },
+    });
+    // Ensure that task is executed after pause completes
+    assert(executionTime >= pauseEnds, 'Queue did not pause for full time.');
+  });
+
+  it('Doesn\'t add delay after pause', async function () {
+    // Create queue to test, with long delay interval
+    const queue = new TaskQueue(40);
+    // Pause queue for 5 ms
+    const start = new Date().getTime();
+    const pauseEnds = new Date(start + 10);
+    queue.pauseUntil(pauseEnds);
+    // Add a task that returns current time to queue
+    const executionTime = await queue.add({
+      task: () => { return Date.now(); },
+    });
+    // Ensure that task is executed after pause completes
+    assert(executionTime >= pauseEnds, 'Queue did not pause for full time.');
+    assert(
+      executionTime < new Date(start + 15),
+      'Queue adds delay after pause'
+    );
+  });
+
+  it('Maintains interval when multiple requests are added', async function () {
+    const queue = new TaskQueue(5);
+    const calls = [];
+    const callTimes = [];
+    // Each task pushes execution time to callTimes
+    for (let i = 0; i < 10; i++) {
+      calls.push(queue.add({
+        task: () => {
+          callTimes.push(new Date().getTime());
+        },
+      }));
+    }
+    await Promise.all(calls);
+    for (let i = 1; i < 10; i++) {
+      assert(
+        callTimes[i] - callTimes[i - 1] >= 5,
+        `Delay interval between calls too short: ${callTimes[i] - callTimes[i - 1]}ms`
+      );
+      assert(
+        callTimes[i] - callTimes[i - 1] <= 8,
+        `Delay interval between calls too long: ${callTimes[i] - callTimes[i - 1]}ms`
+      );
+    }
+  });
+
+  it('Never dequeues multiple tasks at once', async function () {
+    this.skip();
```

Comment:
Is this a test that cannot be written? Or should we discuss it? Or is it just still on your TODO list?

Reply:
I was planning to ask for suggestions on how to test this, since it's hard to engineer this kind of race condition issue